src/dataflow/src/source/mod.rs

```diff
@@ -422,49 +422,126 @@ impl fmt::Debug for SourceMessage {
+    fn inner(&self) -> MzOffset {
+        match self {
+            TimestampUpperBound::Fixed(o) => *o,
+            TimestampUpperBound::Tentative(o) => *o,
+        }
+    }
+
+    fn is_fixed(&self) -> bool {
+        match self {
+            TimestampUpperBound::Fixed(_) => true,
+            TimestampUpperBound::Tentative(_) => false,
+        }
+    }
+}
+
+/// Per-partition consistency information. This effectively acts a cache of
+/// timestamp binding information that needs to be refreshed every time
+/// the source operator is invoked.
+#[derive(Copy, Clone, Debug)]
 struct ConsInfo {
     /// The timestamp we are currently aware of. This timestamp is open iff
-    /// offset < current_upper_bound and closed otherwise.
+    /// offset < current_upper_bound  - 1 (aka there are more offsets we need to find)
+    /// and closed otherwise.
     current_ts: Timestamp,
-    /// Current upper bound for the current timestamp. All offsets <= upper_bound
+    /// Current upper bound for the current timestamp. All offsets < upper_bound
     /// are assigned to current_ts.
-    current_upper_bound: MzOffset,
-    /// the last processed offset
+    current_upper_bound: TimestampUpperBound,
+    /// The last processed offset
     offset: MzOffset,
 }
 
 impl ConsInfo {
     fn new(timestamp: Timestamp) -> Self {
         Self {
             current_ts: timestamp,
-            current_upper_bound: MzOffset { offset: 0 },
+            current_upper_bound: TimestampUpperBound::Fixed(MzOffset { offset: 0 }),
             offset: MzOffset { offset: 0 },
         }
     }
 
-    fn update_timestamp(&mut self, timestamp: Timestamp, upper: MzOffset) {
+    /// Checks if the current binding is sufficient for this offset.
+    ///
+    /// This code assumes we will be ingesting offsets in order.
+    fn is_valid(&self, offset: MzOffset) -> bool {
+        if !self.current_upper_bound.is_fixed() || self.current_upper_bound.inner() > offset {
+            true
+        } else {
+            false
+        }
+    }
+
+    fn update_timestamp_fixed(&mut self, timestamp: Timestamp, upper: MzOffset) {
         assert!(timestamp >= self.current_ts);
-        assert!(upper >= self.current_upper_bound);
+        assert!(upper >= self.current_upper_bound.inner());
 
-        self.current_upper_bound = upper;
+        self.current_upper_bound = TimestampUpperBound::Fixed(upper);
         self.current_ts = timestamp;
     }
 
+    /// Store a new (time, offset) binding in the cache. If no such binding is
+    /// available, we will start proposing our own.
+    fn update_timestamp(&mut self, timestamp: Timestamp, offset: Option<MzOffset>) {
+        if offset.is_some() {
+            self.update_timestamp_fixed(timestamp, offset.expect("known to exist"));
+        } else {
+            assert!(timestamp >= self.current_ts);
+            let old_upper = self.current_upper_bound.inner();
+            self.current_upper_bound = TimestampUpperBound::Tentative(old_upper);
```

Comment:
All this would ever do is to "downgrade" a `Fixed(foo)` to a `Tentative(foo)`, right?

Reply:
this means that lets say at some point in the past we read data up to offset1, and assigned it t1

now we got a new real time timestamp t2, with no stated upper bound of offsets. so we will first finalize the binding of

offsets [0, offset1) -> t1, and tentatively we say that offsets [offset1, offset1) (aka empty interval) are bound to t2. i added some comments to that effect